service/tablet_allocator.cc

```diff
@@ -880,7 +881,165 @@ class load_balancer {
+                // replica of tablet 0 in n4.
+                std::unordered_set<tablet_replica> intersection;
+                std::ranges::set_intersection(r1, r2, std::inserter(intersection, intersection.begin()));
+                const auto [r1_first, r1_last] = std::ranges::remove_if(r1, [&] (tablet_replica r) { return intersection.contains(r); });
+                r1.erase(r1_first, r1_last);
+                const auto [r2_first, r2_last] = std::ranges::remove_if(r2, [&] (tablet_replica r) { return intersection.contains(r); });
+                r2.erase(r2_first, r2_last);
+                // Favor replicas of different tablets that belong to same node. For example:
+                // tablet 0 replicas: [n2:s1, n3:s0]
+                // tablet 1 replicas: [n1:s0, n2:s0]
+                // Replica in n1:s0 cannot follow sibling replica in n2:s1. Otherwise, RF invariant is broken.
+                // Instead, tablet 1 in n2:s0 will be co-located with tablet 0 in n2:s1.
+                std::unordered_map<host_id, tablet_replica> r1_map;
+                std::ranges::transform(r1, std::inserter(r1_map, r1_map.begin()), [] (tablet_replica r) {
+                    return std::make_pair(r.host, r);
+                });
+                for (unsigned i = 0; i < r2.size(); i++) {
+                    auto r1_it = r1_map.find(r2[i].host);
+                    if (r1_it != r1_map.end()) {
+                        return std::make_pair(r1_it->second, r2[i]);
+                    }
+                }
+                // Since sets had intersection subtracted, the remaining replicas are certainly not co-located.
+                if (r1.size() > 0) {
+                    return std::make_pair(r1[0], r2[0]);
+                }
+                return std::nullopt;
+            };
+
+            auto create_migration_info = [] (global_tablet_id gid, tablet_replica src, tablet_replica dst) {
+                auto kind = (src.host != dst.host) ? tablet_transition_kind::migration : tablet_transition_kind::intranode_migration;
+                return tablet_migration_info{kind, gid, src, dst};
+            };
+
+            bool all_colocated = true;
+            co_await tmap.for_each_sibling_tablets([&] (tablet_desc t1, std::optional<tablet_desc> t2_opt) -> future<> {
+                // Be optimistic about migrating tablets, as if they succeeded.
+                // Merge finalization will have to recheck that all sibling tablets are co-located.
+
+                if (!t2_opt) {
+                    on_internal_error(lblogger, format("Unable to find sibling tablet during co-location, with tablet count {}, for table {}",
+                                                       tmap.tablet_count(), table));
+                }
+                auto t2 = *t2_opt;
+
+                auto r1 = get_replicas(t1);
+                auto r2 = get_replicas(t2);
+                if (r1 == r2) {
```

Comment:
don't we need to restrict co-location migrations within a DC?

for example, let's say
DC1{ n1, n3 }
DC2{ n2, n4 }

tablet0 { n1, n4 }
tablet1 { n2, n3 }

we cannot migrate t1:n2 into t0:n1, since they're on different DCs.

if the problem is that nodes is not a good source of truth, I think I can instead rely on topology that can tell me if a particular host id lives in the target DC.

the decision should still take a look at skip list to avoid emitting a co-location migraton if either the source or destination is down.

Reply:
I am not exactly sure what you meant by "This should be computed not in per-DC plan, but the global plan.". Maybe you still want the DC aware co-location (which is what you proposed in the original design), but want a second pass over the replica sets when determining that everything is co-located?

Isn't it enough if I fix get_replicas() to not discard any replica within the DC? then I can combine all_colocated from all resize plans (as happens in the current version).

Reply:
> I am not exactly sure what you meant by "This should be computed not in per-DC plan, but the global plan.". Maybe you still want the DC aware co-location (which is what you proposed in the original design), but want a second pass over the replica sets when determining that everything is co-located?

Yes, I only meant that all_colocated should be moved to the global plan making. Co locating migration scheduling should be per DC. 

> 
> Isn't it enough if I fix get_replicas() to not discard any replica within the DC? then I can combine all_colocated from all resize plans (as happens in the current version).

It seems more complicated than necessary to maintain that state. It's also potentially unsafe because it relies on the assumption that we invoked plan maker for all DCs present in the replica list. I'm not sure it will always hold in the future.